File: go/src/complex/complex.go
package complex

import (
    "fmt"
    "math"
)

type Complex interface {
    GetReal() float64
    GetImage() float64
    GetAbs() float64
    GetArg() float64
    ToString() string
}

func Add(a Complex, b Complex) Complex {
    return CreateRect(a.GetReal() + b.GetReal(), 
        a.GetImage() + b.GetImage())
}

type ComplexRect struct {
    real, image float64
}

func CreateRect(r float64, i float64) ComplexRect {
    return ComplexRect{r, i}
}

func (c ComplexRect) GetReal() float64 {
    return c.real
}

func (c ComplexRect) GetImage() float64 {
    return c.image
}

func (c ComplexRect) GetAbs() float64 {
    return math.Sqrt(c.real * c.real + c.image * c.image)
}

func (c ComplexRect) GetArg() float64 {
    return math.Atan(c.image / c.real)
}

func (c ComplexRect) ToString() string {
    var sgn = "+"
    if (c.image < 0) {
        sgn = ""
    }
    return fmt.Sprintf("%f%s%fi", c.real, sgn, c.image)
}

type ComplexPolar struct {
    abs, arg float64
}

func CreatePolar(r float64, a float64) ComplexPolar {
    return ComplexPolar{r, a}
}

func (c ComplexPolar) GetReal() float64 {
    return c.abs * math.Cos(c.arg)
}

func (c ComplexPolar) GetImage() float64 {
    return c.abs * math.Sin(c.arg)
}

func (c ComplexPolar) GetAbs() float64 {
    return c.abs
}

func (c ComplexPolar) GetArg() float64 {
    return c.arg
}

func (c ComplexPolar) ToString() string {
    return fmt.Sprintf("%f*exp(%fi)", c.abs, c.arg)
}

File: clj/complex.clj
(ns complex
  (:require [dynbind]))

(def rect-tag 'rect)
(def polar-tag 'polar)

(defn make-rect [real image]
  (dynbind/attach-tag rect-tag (list (float real) (float image)))
)

(defn make-polar [abs arg]
  (dynbind/attach-tag polar-tag (list (float abs) (float arg)))
)

(dynbind/interface get-real)
(dynbind/interface get-image)
(dynbind/interface get-abs)
(dynbind/interface get-arg)
(dynbind/interface to-string)
; (println (macroexpand '(dynbind/interface to-string)))
; (println to-string)

(defn install-rect []
  (letfn [(get-real [c]
            (first c))
          (get-image [c]
            (first (rest c)))
          (get-abs [c]
            (let [r (get-real c) i (get-image c)]
              (Math/sqrt (+ (* r r) (* i i)))))
          (get-arg [c]
            (let [r (get-real c) i (get-image c)]
              (Math/atan (/ i r))))
          (to-string [c]
            ; (println c)
            (let [r (get-real c)
                  i (get-image c)
                  sgn-i (if (< i 0) "" "+")]
              (format "%f%s%fi" r sgn-i i)))]

    (dynbind/install rect-tag 'get-real get-real)
    (dynbind/install rect-tag 'get-image get-image)
    (dynbind/install rect-tag 'get-abs get-abs)
    (dynbind/install rect-tag 'get-arg get-arg)
    (dynbind/install rect-tag 'to-string to-string)
  )
)

(defn install-polar []
  (letfn [(get-real [c]
            (* (get-abs c) (Math/cos (get-arg c))))
          (get-image [c]
            (* (get-abs c) (Math/sin (get-arg c))))
          (get-abs [c]
            (first c))
          (get-arg [c]
            (first (rest c)))
          (to-string [c]
            ; (println c)
            (let [r (get-abs c)
                  i (get-arg c)]
              (format "%f*exp(%fi)" r i)))]

    (dynbind/install polar-tag 'get-real get-real)
    (dynbind/install polar-tag 'get-image get-image)
    (dynbind/install polar-tag 'get-abs get-abs)
    (dynbind/install polar-tag 'get-arg get-arg)
    (dynbind/install polar-tag 'to-string to-string)
  )
)

(install-rect)
(install-polar)

; (println (deref dynbind/vtable))

(defn add [c d]
  (make-rect (+ (get-real c) (get-real d))
             (+ (get-image c) (get-image d))
  )
)
